### PR TITLE
Flink: make `StatisticsOrRecord` to be correctly serialized and deser…

### PR DESCRIPTION
**Pull request by huyuanfeng2018**
## background
  When I configure distribution-mode=RANGE in the flink task, the task can process the data as expected, but the processing speed is insufficient. Like #7393, there is a serialization problem.

image

## Change part
* Add `StatisticsOrRecordTypeInformation` and specify it as `outTypeInfo`
* add test  `org.apache.iceberg.flink.sink.shuffle.TestStatisticsOrRecordTypeInformation`

### CODE REVIEW

**Comment by huyuanfeng2018** [comment]
@stevenzwu can you review it for me, thank you very much !

**Comment by huyuanfeng2018** [comment]
> @stevenzwu @pvary can you review it for me, thank you very much !

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/StatisticsOrRecordTypeInformation.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+          && that.rowTypeInformation.equals(this.rowTypeInformation)
+          && that.globalStatisticsSerializer.equals(this.globalStatisticsSerializer);
+    } else {
+      return false;
+    }
+  }
+
+  @Override
+  public int hashCode() {
+    return 31 * rowTypeInformation.hashCode()
```
Comment (stevenzwu) [inline flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/StatisticsOrRecordTypeInformation.java:107]: let's use `Objects.hash(...)`

**Review by stevenzwu — COMMENTED** [review]
@huyuanfeng2018 thx for catching and fixing the bug.

Do you have flame graph or numbers that compare the before and after fix?

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/StatisticsOrRecordTypeInformation.java (hunk 1 above)
Comment (huyuanfeng2018) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/StatisticsOrRecordTypeInformation.java]: fixed

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/StatisticsOrRecordTypeInformation.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    return "StatisticsOrRecord";
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    } else if (o != null && this.getClass() == o.getClass()) {
+      StatisticsOrRecordTypeInformation that = (StatisticsOrRecordTypeInformation) o;
+      return that.sortOrder.equals(this.sortOrder)
```
Comment (stevenzwu) [inline flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/StatisticsOrRecordTypeInformation.java:98]: nit: Iceberg style only uses `this.` when updating/writing the value (not for read).

**Review by stevenzwu — APPROVED** [review]
Just nit comment on style

**Comment by huyuanfeng2018** [comment]
> Just a nit comment on style

Thanks for review, fixed.

**Review by stevenzwu — APPROVED** [review]

**Comment by stevenzwu** [comment]
thanks @huyuanfeng2018 . can you create the back port PR too?

**Comment by huyuanfeng2018** [comment]
> thanks @huyuanfeng2018 . can you create the back port PR too?

ok, I will be opening PR later this week~

**Pull request merged**